src/CodeStyle/Tools/Program.cs

```diff
@@ -0,0 +1,302 @@
+
+                        bool AddRule(DiagnosticDescriptor rule)
+                        {
+                            var (isEnabledByDefault, severity) = GetEnabledByDefaultAndSeverity(rule, analysisMode);
+                            if (rule.IsEnabledByDefault == isEnabledByDefault &&
+                                severity == rule.DefaultSeverity)
+                            {
+                                // Rule had the same default severity and enabled state in the release.
+                                // We do not need to generate any entry.
+                                return false;
+                            }
+
+                            var severityString = GetRuleSeverity(isEnabledByDefault, severity);
+
+                            result.AppendLine();
+                            result.AppendLine($"# {rule.Id}: {rule.Title}");
+                            result.AppendLine($@"dotnet_diagnostic.{rule.Id}.severity = {severityString}");
+                            return true;
+                        }
+
+                        (bool isEnabledByDefault, DiagnosticSeverity effectiveSeverity) GetEnabledByDefaultAndSeverity(DiagnosticDescriptor rule, AnalysisMode analysisMode)
+                        {
+                            bool isEnabledInNonDefaultMode;
+                            switch (analysisMode)
+                            {
+                                case AnalysisMode.None:
+                                    // Disable all rules by default.
+                                    return (isEnabledByDefault: false, DiagnosticSeverity.Warning);
+
+                                case AnalysisMode.All:
+                                    // Escalate all rules which can be enabled on build.
+                                    isEnabledInNonDefaultMode = !rule.CustomTags.Contains(s_neverTag);
+                                    break;
+
+                                case AnalysisMode.Minimum:
+                                    // Escalate all highly recommended rules.
+                                    isEnabledInNonDefaultMode = rule.CustomTags.Contains(s_highlyRecommendedTag);
+                                    break;
+
+                                case AnalysisMode.Recommended:
+                                    // Escalate all recommended and highly recommended rules.
+                                    isEnabledInNonDefaultMode = rule.CustomTags.Contains(s_highlyRecommendedTag) || rule.CustomTags.Contains(s_recommendedTag);
+                                    break;
+
+                                case AnalysisMode.Default:
+                                    // Retain the default severity and enabled by default values.
+                                    isEnabledInNonDefaultMode = false;
+                                    break;
+
+                                default:
+                                    throw new NotSupportedException();
```

Comment:
Adding some message would help to understand what's going-on if the issue does happen.